### DISCUSSION TITLE
import plugin as a way using bundle but expects as peer dependency to not add into bundled script

### DISCUSSION THREAD

**Discussion by shashank-brightness** [alpinejs/alpine#4235] [1. Help] [answered]
Hi i am running into one weird issue, where i need to declare plugin in my package as `peerDependency` and expects the consumer to have this plugin installed but not finding the way around it.

As this is related to webpack externals, which is found [here](https://webpack.js.org/configuration/externals/), i followed and added my library to be installed as umd without needing to import it anywhere.

```
externals: [
    {
      'bootstrap/js/dist/dropdown': 'bootstrap.Dropdown',
      'bootstrap/js/dist/tab': 'bootstrap.Tab',
      'bootstrap/js/dist/tooltip': 'bootstrap.Tooltip',
      'bootstrap/js/dist/modal': 'bootstrap.Modal',
      'alpinejs': 'Alpine',
      '@alpinejs/intersect': 'Alpine.$intersect',
      
    }
  ],
```

and this is how my `package.json` looks for `peerDependency` and as i did not want to bundle intersect plugin with my own package i declared it as peer deps and it is expected to be available in consumers enviornment as global variable.

```
"peerDependencies": {
    "@alpinejs/intersect": "^3.14.0",
    "@alpinejs/persist": "^3.13.3",
    "alpinejs": "^3.13.2",
    "bootstrap": "^5.1.3",
}
```

and this is how webpack converts import statements for that thing to `window.thatthing` and then uses in the script assuming that it is available in user's end. 

i.e. when i write `const Alpine = require('alpinejs')` in any module, it takes it from window level because alpinejs module is mapped to that variable which should be at window level.

However i am not getting how i should do this when plugin syntax is there in script because if i add plugin from script directly before alpine, it would raise error about plugin being registered twice when `Alpine.plugin(Alpine.$intersect)` runs after `import intersect from '@alpinejs/intersect'` is converted to `window.Alpine.$intersect`.

**Comment by shashank-brightness** [comment]
I also noted that strategy of defining externals is not working at all with `alpinejs` because when we do `import Alpine from 'alpinejs';`, it converts import to `alpinejs` module to `window.Alpine` (and tries assigning that to constant named `Alpine`) which unfortunately does not exist at that time as my script gets imported before alpine and not after it.

**Comment by ekwoka** [comment]
It's unclear to me what you are trying to do.

why can't you just import Alpine like normal?

**Reply by shashank-brightness** [reply]
I want to convert all import statements in bundle to be delegated to consumer side because i don';t want to bundle each and every  library which my package is dependent on along with my own npm package because they are externals. 

for example when i say `import bootstrap from 'bootstrap'`, at this time variable named `bootstrap` is assigned to whatever available at `window.bootstrap` when bootstrap is in the externals list of webpack and so that webpack knows that this will be available in browser enviornment and does not bundle entire bootstrap library with my code (thus saving file size) because that's what mapped in externals config of `webpack.config.js`

**Reply by shashank-brightness** [reply]
without the `bootstrap` entry in `externals` configuration of webpack, it would have taken 'bootstrap' module from `node_modules/bootstrap` and not from `window.bootstrap` and also have added entire bootstrap library as part of my final bundled script which i want to use as umd library.

**Reply by shashank-brightness** [reply]
If i don;t import any of alpine js code then it works but import statements imo should be convertible to the global way while using webpack `externals` so that when developer starts developing package, he adds import/require statements as normal and if later needed then he can add entry in the externals configuration if it is known that library is already present at consumer side and he does not need to import it again.

**Reply by ekwoka** [reply]
So you just assume those other things will be available? or what? It's a bit unclear to me...

What if the user doesn't also have those assigned on the window?

> while using webpack externals

Well, it's outdated systems, so finding support is unlikely.

But yeah, this is really a webpack externals and your setup problem. You are trying ot tell webpack window.Alpine will exist, but you also know it won't exist. You have to choose one.

So there isn't really anything Alpine can do about that.

**Reply by shashank-brightness** [reply]
I somewhat agree that this is not the alpine issue but i somewhat also believe that it is not of webpack as well.
It does not necessarily available on window but atleast the goal is that **import/require should be convertible to the already existing things in browser if dev later got to to know that those were already added somewhere** and so don't endup adding twice. 
If some other place has not that at window level? no problem, then i will remove externals config so it will be added from `node_modules`. So much flexibility here.
Window is used as global in browser so webpack (and I) also mentioned at window level.

Imo, webpack is not the actual issue here because it provides wide range of options when this issue is encountered. i.e. it also provides wide range of options to export library via amd, es6 modules, umd and so many other options in its `externals` config. i choose window because in my project it was necessary to do so and i can't include all sorts of other things in library which would make it more heavier.
e.g. i am even using `i18Next` library for translation. 

please lmk if you have any other suggestion for this scenario when `mypackage` has imported **something** via ES6 import/require and later that **something** was found to be already existing via script adding way at someplace but not at some other place. 

Do you think i can just provide documentation straight to the users like **hey this (alpine here) is required to be added if you use mypackage** thus forgetting entirely about the idea of converting from import/require?

**Reply by ekwoka** [reply, marked as answer]
Well, you have to decide the order you want things loaded.

Realistically, no reason not to do normal bundling (maybe investigate vite as well).

Additionally, dynamic imports may work with externals jsut fine, so dynamically import alpine once you get the alpine event only.

Or export your thing only as a plugin that consumes Alpine, so it never needs to import it at all.
